File: examples/grids/typescript/grid.ts
/**
 * Output:
 * ---------⌝
 * ----------
 * ----?????-
 * ----------
 * ----------
 * --!!!-----
 * --!!!-----
 * ----------
 * ----------
 * ⌞---------
 */
const demo = () => {
  const n = 10;

  // Using a simple object is fine. A map doesn't add much functionality.
  const grid: Record<string, string | undefined> = {};

  // Assignment must use coordinate strings:
  grid[coord(0, 0)] = "⌞";
  grid[coord(n - 1, n - 1)] = "⌝";

  fill(grid, "!", [2, 3], [5, 5]);
  fill(grid, "?", [4, 7], [9, 8]);

  console.log(stringify(grid, n));
};

/**
 * Only strings work as object keys, and they're easy to debug.
 */
const coord = (x: number, y: number): string => `(${x}, ${y})`;

/**
 * You can use tuple arrays in typescript, but they'll have to be
 * converted to strings for any look-up.
 */
const fill = (
  grid: Record<string, string | undefined>,
  value: string,
  start: [number, number],
  stop: [number, number]
) => {
  for (let x = start[0]; x < stop[0]; x += 1) {
    for (let y = start[1]; y < stop[1]; y += 1) {
      grid[coord(x, y)] = value;
    }
  }
};

/**
 * Stringify with (0, 0) in the lower-left corner.
 */
const stringify = (
  grid: Record<string, string | undefined>,
  n: number
): string => {
  const rows = [];
  for (let y = n - 1; 0 <= y; y -= 1) {
    const row = [];
    for (let x = 0; x < n; x += 1) {
      row.push(coord(x, y) in grid ? grid[coord(x, y)] : "-");
    }

    rows.push(row);
  }

  return rows.map((row) => row.join("")).join("\n");
};

demo();

File: examples/grids/ruby/grid.rb
#
def demo()
    n = 10

    grid = {}

    coord = [1, 1]
    grid[coord] = "a"
    coord[0] = 2
    grid[coord] = "b"
    coord[0] = 1
    grid[coord] = "a"


    grid[[0, 0]] = "⌞"
    grid[[n - 1, n - 1]] = "⌝"

    print(stringify(grid, n))
end

def stringify(grid, n)
    rows = []
    (0...n).reverse_each do |y|
        row = []
        (0...n).each do |x|
            value = grid.fetch([x, y], "-")
            row.push(value)
        end

        rows.push(row)
    end

    rows.map { |row| row.join("") }.join("\n")
end

demo